src/Features/Core/Portable/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -152,5 +176,51 @@ static bool CanConvertSectionForSwitchExpression(bool supportsOrPattern, Analyze
                 return supportsOrPattern && section.Labels.All(label => label.Guards.IsDefaultOrEmpty);
             }
         }
+
+        protected sealed override async Task FixAllAsync(
+            Document document,
+            ImmutableArray<TextSpan> fixAllSpans,
+            SyntaxEditor editor,
+            CodeActionOptionsProvider optionsProvider,
+            string? equivalenceKey,
+            CancellationToken cancellationToken)
+        {
+            var convertToSwitchExpression = equivalenceKey == SwitchExpressionEquivalenceKey;
+            var syntaxFactsService = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            // Get all the descendant if statements to process.
+            var ifStatements = editor.OriginalRoot.DescendantNodes().OfType<TIfStatementSyntax>();
+
+            // We're going to be continually editing this tree. Track all the nodes we
+            // care about so we can find them across each edit.
+            document = document.WithSyntaxRoot(editor.OriginalRoot.TrackNodes(ifStatements));
+
+            // Process the if statements.
+            foreach (var originalIfStatement in ifStatements)
+            {
+                // Only process if statements fully within a fixAllSpan
+                if (!fixAllSpans.Any(fixAllSpan => fixAllSpan.Contains(originalIfStatement.Span)))
+                    continue;
+
+                // Get current root, if statement and semantic model.
+                var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var ifStatement = root.GetCurrentNodes(originalIfStatement).SingleOrDefault();
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+                // Check if the refactoring is applicable for this if statement.
+                if (!ShouldOfferRefactoring(ifStatement, semanticModel, syntaxFactsService, out var analyzer, out var sections, out var target))
+                    continue;
+
+                // When converting to switch expression, we need to perform an additional check to ensure this conversion is possible.
+                if (convertToSwitchExpression && !CanConvertToSwitchExpression(analyzer.Supports(Feature.OrPattern), sections))
+                    continue;
+
+                document = await UpdateDocumentAsync(document, target, ifStatement, sections,
+                    analyzer.Features, convertToSwitchExpression, cancellationToken).ConfigureAwait(false);
+            }
+
+            var updatedRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            editor.ReplaceNode(editor.OriginalRoot, updatedRoot);
```

Comment:
so, overall this algorithm seems fine to me (and likely the most correct).  But it's also a bit expensive as it constantly forks.  I'm curious if there was a reason you didn't take this approach (which we commonly do elsewhere):

1. get teh if statements
2. process them in reverse order.
3. get the resultant rewritten node for it.
4. do a replacement of it using the Editor.  Use the Replacement function that gets passed the currently transformed node along.  

Note: to make this work, you'd need to chnage UpdateDocumentAsync slightly.  Instead of returning the brand new document, it would just give you the switch to replace the `if-statement` with.  

Then, we're single-pass, and we can reuse all semantics.

Reply:
I did try the approach you mentioned upfront, but that doesn't work, in fact it fails even for the FixAll tests I have added.

1. Processing statements in reverse order means nested if statements get converted to switch first before the outer ones. This is not desirable as this refactoring can actually fold the nested if into outer switch. For example, for the fix all test added, we have following as input:

```csharp
        if (i == 3)
        {
            return 0;
        }
        else
        {
            if (i == 6) return 1;
            if (i == 7) return 1;
            return 0;
        }
```

and processing the outer if leads the refactoring to generate the following:

```csharp
        return i switch
        {
            3 => 0,
            6 => 1,
            7 => 1,
            _ => 0
        };
```

So, the inner if statements have also been folded into the switch expression generated when outermost if statement is processed. This behavior also means that some of the nested if statements identified at start of the FixAll operation might not exist when we eventually reach that node for fixing, so this forces us to operate on each if statement at a time, and updating the root, document and semantic model after each if is converted to switch.

I'll add a detailed comment about this in code for clarity.

Reply:
that makes perfect sense.  thanks for the explanation.  def good to have in the code :)

Reply:
Added with https://github.com/dotnet/roslyn/pull/61874/commits/eb5c3f3408aae2aaab56192b57e36f0812fb11cd